File: php/src/Protocols/Mpp/Core/Blake3.php
<?php

declare(strict_types=1);

namespace PayKit\Protocols\Mpp\Core;

/**
 * Pure-PHP BLAKE3 (default 256-bit output, no key, no derive-key).
 *
 * The payment-channels program commits split recipients with a BLAKE3 hash of
 * the distribution preimage (`len(u32 LE) || (recipient(32) || bps(u16 LE))*`).
 * The Rust spine derives that hash with the `blake3` crate
 * (`rust/crates/mpp/src/program/payment_channels.rs::distribution_hash`); PHP
 * has no native BLAKE3 in `hash_algos()`, so this implements the unkeyed hash
 * directly from the BLAKE3 specification.
 *
 * Only the hashing surface needed by the session intent is implemented:
 * arbitrary-length input, fixed 32-byte output, no keyed/KDF modes and no
 * extendable-output (XOF) beyond 32 bytes. Everything is 32-bit-word math kept
 * inside PHP integers with explicit `& 0xFFFFFFFF` masking so it is correct on
 * both 32-bit and 64-bit builds.
 *
 * @see https://github.com/BLAKE3-team/BLAKE3-specs BLAKE3 specification
 */
final class Blake3
{
    private const OUT_LEN = 32;
    private const BLOCK_LEN = 64;
    private const CHUNK_LEN = 1024;

    private const CHUNK_START = 1 << 0;
    private const CHUNK_END = 1 << 1;
    private const PARENT = 1 << 2;
    private const ROOT = 1 << 3;

    /** @var list<int> */
    private const IV = [
        0x6A09E667, 0xBB67AE85, 0x3C6EF372, 0xA54FF53A,
        0x510E527F, 0x9B05688C, 0x1F83D9AB, 0x5BE0CD19,
    ];

    /** @var list<int> */
    private const MSG_PERMUTATION = [2, 6, 3, 10, 7, 0, 4, 13, 1, 11, 12, 5, 9, 14, 15, 8];

    private function __construct()
    {
    }

    /**
     * Compute the 32-byte BLAKE3 hash of the input bytes.
     */
    public static function hash(string $input): string
    {
        $chainingValue = self::IV;
        $len = strlen($input);

        // Single-chunk fast path covers the entire session distribution
        // preimage (under 1024 bytes) but the chunk-tree code below handles
        // arbitrary input for parity with the Rust reference.
        if ($len <= self::CHUNK_LEN) {
            $cv = self::hashChunk($input, 0, self::ROOT);
            return self::wordsToBytes($cv);
        }

        // General case: hash every 1024-byte chunk into a chaining value, then
        // combine them into a balanced binary tree. The root node's compression
        // carries the ROOT flag, so it is computed last and separately. Session
        // distribution preimages never reach this path (they fit one chunk),
        // but it is implemented for full parity with the Rust reference.
        /** @var list<list<int>> $chunkCvs */
        $chunkCvs = [];
        $chunkCounter = 0;
        $offset = 0;
        while ($offset < $len) {
            $chunk = substr($input, $offset, self::CHUNK_LEN);
            $offset += self::CHUNK_LEN;
            $chunkCvs[] = self::hashChunk($chunk, $chunkCounter, 0);
            $chunkCounter += 1;
        }

        return self::wordsToBytes(self::mergeRoot($chunkCvs, 0, count($chunkCvs)));
    }

    /**
     * Merge chunk chaining values [$start, $end) into a single node. The
     * top-level call returns the ROOT output words; interior nodes return
     * chaining values. The left subtree always covers the largest power of two
     * strictly less than the span, matching the BLAKE3 tree layout.
     *
     * @param list<list<int>> $chunkCvs
     * @return list<int>
     */
    private static function mergeRoot(array $chunkCvs, int $start, int $end): array
    {
        return self::mergeNode($chunkCvs, $start, $end, self::ROOT);
    }

    /**
     * @param list<list<int>> $chunkCvs
     * @return list<int>
     */
    private static function mergeNode(array $chunkCvs, int $start, int $end, int $rootFlag): array
    {
        $count = $end - $start;
        if ($count === 1) {
            return $chunkCvs[$start];
        }
        $leftSpan = 1;
        while ($leftSpan * 2 < $count) {
            $leftSpan *= 2;
        }
        $mid = $start + $leftSpan;
        $left = self::mergeNode($chunkCvs, $start, $mid, 0);
        $right = self::mergeNode($chunkCvs, $mid, $end, 0);
        return self::parentCv($left, $right, self::IV, $rootFlag);
    }

    /**
     * Hash one chunk (up to 1024 bytes) and return its 8-word chaining value
     * (or root output words when ROOT is set in $extraFlags).
     *
     * @return list<int>
     */
    private static function hashChunk(string $chunk, int $chunkCounter, int $extraFlags): array
    {
        $cv = self::IV;
        $len = strlen($chunk);
        $blockCount = (int) ceil(max($len, 1) / self::BLOCK_LEN);

        for ($i = 0; $i < $blockCount; $i++) {
            $block = substr($chunk, $i * self::BLOCK_LEN, self::BLOCK_LEN);
            $blockLen = strlen($block);
            $flags = 0;
            if ($i === 0) {
                $flags |= self::CHUNK_START;
            }
            if ($i === $blockCount - 1) {
                $flags |= self::CHUNK_END | $extraFlags;
            }
            $block = str_pad($block, self::BLOCK_LEN, "\x00");
            $words = self::compress($cv, $block, $chunkCounter, $blockLen, $flags);
            // Chaining value is the first 8 output words.
            $cv = array_slice($words, 0, 8);
            if (($flags & self::ROOT) !== 0) {
                return $cv;
            }
        }

        return $cv;
    }

    /**
     * Compute a parent node chaining value from two child chaining values.
     *
     * @param list<int> $leftCv
     * @param list<int> $rightCv
     * @param list<int> $key
     * @return list<int>
     */
    private static function parentCv(array $leftCv, array $rightCv, array $key, int $flags): array
    {
        $blockWords = array_merge($leftCv, $rightCv);
        $block = '';
        foreach ($blockWords as $word) {
            $block .= pack('V', $word & 0xFFFFFFFF);
        }
        $words = self::compress($key, $block, 0, self::BLOCK_LEN, self::PARENT | $flags);
        return array_slice($words, 0, 8);
    }

    /**
     * BLAKE3 compression function. Returns 16 output words.
     *
     * @param list<int> $chainingValue 8 words
     * @return list<int>
     */
    private static function compress(
        array $chainingValue,
        string $block,
        int $counter,
        int $blockLen,
        int $flags
    ): array {
        $m = array_values(unpack('V16', $block));

        $counterLow = $counter & 0xFFFFFFFF;
        $counterHigh = ($counter >> 32) & 0xFFFFFFFF;

        $state = [
            $chainingValue[0], $chainingValue[1], $chainingValue[2], $chainingValue[3],
            $chainingValue[4], $chainingValue[5], $chainingValue[6], $chainingValue[7],
            self::IV[0], self::IV[1], self::IV[2], self::IV[3],
            $counterLow, $counterHigh, $blockLen & 0xFFFFFFFF, $flags & 0xFFFFFFFF,
        ];

        for ($round = 0; $round < 7; $round++) {
            self::round($state, $m);
            $m = self::permute($m);
        }

        for ($i = 0; $i < 8; $i++) {
            $state[$i] = ($state[$i] ^ $state[$i + 8]) & 0xFFFFFFFF;
            $state[$i + 8] = ($state[$i + 8] ^ $chainingValue[$i]) & 0xFFFFFFFF;
        }

        return $state;
    }

    /**
     * @param array<int,int> $state 16 words (modified in place)
     * @param array<int,int> $m 16 message words
     */
    private static function round(array &$state, array $m): void
    {
        // Columns
        self::g($state, 0, 4, 8, 12, $m[0], $m[1]);
        self::g($state, 1, 5, 9, 13, $m[2], $m[3]);
        self::g($state, 2, 6, 10, 14, $m[4], $m[5]);
        self::g($state, 3, 7, 11, 15, $m[6], $m[7]);
        // Diagonals
        self::g($state, 0, 5, 10, 15, $m[8], $m[9]);
        self::g($state, 1, 6, 11, 12, $m[10], $m[11]);
        self::g($state, 2, 7, 8, 13, $m[12], $m[13]);
        self::g($state, 3, 4, 9, 14, $m[14], $m[15]);
    }

    /**
     * @param array<int,int> $state modified in place
     */
    private static function g(array &$state, int $a, int $b, int $c, int $d, int $mx, int $my): void
    {
        $state[$a] = ($state[$a] + $state[$b] + $mx) & 0xFFFFFFFF;
        $state[$d] = self::rotr($state[$d] ^ $state[$a], 16);
        $state[$c] = ($state[$c] + $state[$d]) & 0xFFFFFFFF;
        $state[$b] = self::rotr($state[$b] ^ $state[$c], 12);
        $state[$a] = ($state[$a] + $state[$b] + $my) & 0xFFFFFFFF;
        $state[$d] = self::rotr($state[$d] ^ $state[$a], 8);
        $state[$c] = ($state[$c] + $state[$d]) & 0xFFFFFFFF;
        $state[$b] = self::rotr($state[$b] ^ $state[$c], 7);
    }

    private static function rotr(int $value, int $bits): int
    {
        $value &= 0xFFFFFFFF;
        return (($value >> $bits) | ($value << (32 - $bits))) & 0xFFFFFFFF;
    }

    /**
     * @param list<int> $m
     * @return list<int>
     */
    private static function permute(array $m): array
    {
        $out = [];
        foreach (self::MSG_PERMUTATION as $index) {
            $out[] = $m[$index];
        }
        return $out;
    }

    /**
     * @param list<int> $words 8 words
     */
    private static function wordsToBytes(array $words): string
    {
        $out = '';
        for ($i = 0; $i < self::OUT_LEN / 4; $i++) {
            $out .= pack('V', $words[$i] & 0xFFFFFFFF);
        }
        return $out;
    }
}

File: lua/pay_kit/solana/blake3.lua
--[[
Pure-LuaJIT BLAKE3 (hash mode, no keyed/derive variants).

The MPP payment-channels program commits a 32-byte distribution hash at
channel open time, computed with BLAKE3 over the recipient preimage. None
of the other PayKit crypto primitives (`pay_kit.util._mpp_crypto`) carry
BLAKE3, so the session intent ships its own implementation here.

This is a faithful port of the BLAKE3 reference (the official spec at
<https://github.com/BLAKE3-team/BLAKE3-specs>). It implements the regular
unkeyed hash with the default 32-byte output, which is all the
distribution-hash preimage needs. The preimage the payment-channels
program hashes is small (`u32 len || (pubkey32 || u16 bps)*`), so the
single-threaded, single-output reference structure is sufficient; there is
no need for the SIMD / multi-chunk-tree fast paths.

LuaJIT ships the `bit` library natively, so the 32-bit word ops use it
directly rather than the pure-Lua `pay_kit.util.bit` fallback.
]]

local bit = require('bit')
local band, bxor = bit.band, bit.bxor
local rshift, ror = bit.rshift, bit.ror
local tobit, bnot = bit.tobit, bit.bnot

local M = {}

local OUT_LEN = 32
local BLOCK_LEN = 64
local CHUNK_LEN = 1024

-- Domain-separation flags.
local CHUNK_START = 1
local CHUNK_END = 2
local ROOT = 8

local IV = {
  0x6A09E667, 0xBB67AE85, 0x3C6EF372, 0xA54FF53A,
  0x510E527F, 0x9B05688C, 0x1F83D9AB, 0x5BE0CD19,
}

-- Message permutation applied between rounds. These are the canonical
-- 0-based BLAKE3 permutation indices [2,6,3,10,7,0,4,13,1,11,12,5,9,14,15,8]
-- shifted by +1 so they index a 1-based 16-word message table.
local MSG_PERMUTATION = {
  3, 7, 4, 11, 8, 1, 5, 14, 2, 12, 13, 6, 10, 15, 16, 9,
}

-- 32-bit add that wraps; `bit.tobit` normalizes to a signed 32-bit int but
-- equality/serialization treat the low 32 bits, which is what we want.
local function add(a, b)
  return tobit(a + b)
end

local function g(state, a, b, c, d, mx, my)
  state[a] = add(add(state[a], state[b]), mx)
  state[d] = ror(bxor(state[d], state[a]), 16)
  state[c] = add(state[c], state[d])
  state[b] = ror(bxor(state[b], state[c]), 12)
  state[a] = add(add(state[a], state[b]), my)
  state[d] = ror(bxor(state[d], state[a]), 8)
  state[c] = add(state[c], state[d])
  state[b] = ror(bxor(state[b], state[c]), 7)
end

local function round(state, m)
  -- Columns.
  g(state, 1, 5, 9, 13, m[1], m[2])
  g(state, 2, 6, 10, 14, m[3], m[4])
  g(state, 3, 7, 11, 15, m[5], m[6])
  g(state, 4, 8, 12, 16, m[7], m[8])
  -- Diagonals.
  g(state, 1, 6, 11, 16, m[9], m[10])
  g(state, 2, 7, 12, 13, m[11], m[12])
  g(state, 3, 8, 9, 14, m[13], m[14])
  g(state, 4, 5, 10, 15, m[15], m[16])
end

local function permute(m)
  local out = {}
  for i = 1, 16 do
    out[i] = m[MSG_PERMUTATION[i]]
  end
  return out
end

-- The core compression function. `chaining` is 8 words, `block_words` is
-- 16 words, returns 16 output words (only the first 8 are used for the
-- chaining value; the full 16 are used for the root output).
local function compress(chaining, block_words, counter, block_len, flags)
  local counter_low = tobit(counter % 4294967296)
  local counter_high = tobit(math.floor(counter / 4294967296) % 4294967296)
  local state = {
    chaining[1], chaining[2], chaining[3], chaining[4],
    chaining[5], chaining[6], chaining[7], chaining[8],
    IV[1], IV[2], IV[3], IV[4],
    counter_low, counter_high, tobit(block_len), tobit(flags),
  }
  local m = block_words
  for _ = 1, 6 do
    round(state, m)
    m = permute(m)
  end
  round(state, m)

  local out = {}
  for i = 1, 8 do
    out[i] = bxor(state[i], state[i + 8])
    out[i + 8] = bxor(state[i + 8], chaining[i])
  end
  return out
end

-- Read 16 little-endian 32-bit words from a 64-byte block string. The block
-- is zero-padded to 64 bytes by the caller.
local function words_from_block(block)
  local words = {}
  for i = 0, 15 do
    local b0 = block:byte(i * 4 + 1) or 0
    local b1 = block:byte(i * 4 + 2) or 0
    local b2 = block:byte(i * 4 + 3) or 0
    local b3 = block:byte(i * 4 + 4) or 0
    words[i + 1] = tobit(b0 + b1 * 256 + b2 * 65536 + b3 * 16777216)
  end
  return words
end

-- Compute the BLAKE3 hash of `input` (a byte string) and return a 32-byte
-- raw digest string.
--
-- This handles inputs of any length by splitting into 1024-byte chunks and
-- combining their chaining values up the binary tree, exactly as the
-- reference describes. The distribution-hash preimage is far below one
-- chunk, but the full path keeps the implementation correct for any input.
function M.hash(input)
  input = input or ''

  -- Build the list of chunk chaining values.
  local chunk_cvs = {}
  local total = #input
  local num_chunks = math.max(1, math.ceil(total / CHUNK_LEN))

  for chunk_index = 0, num_chunks - 1 do
    local chunk_start = chunk_index * CHUNK_LEN
    local chunk_end = math.min(chunk_start + CHUNK_LEN, total)
    local chunk = input:sub(chunk_start + 1, chunk_end)
    local chunk_data_len = #chunk
    local num_blocks = math.max(1, math.ceil(chunk_data_len / BLOCK_LEN))

    local cv = { IV[1], IV[2], IV[3], IV[4], IV[5], IV[6], IV[7], IV[8] }
    for block_index = 0, num_blocks - 1 do
      local bstart = block_index * BLOCK_LEN
      local bend = math.min(bstart + BLOCK_LEN, chunk_data_len)
      local block = chunk:sub(bstart + 1, bend)
      local block_len = #block
      -- Zero-pad to 64 bytes.
      if block_len < BLOCK_LEN then
        block = block .. string.rep('\0', BLOCK_LEN - block_len)
      end

      local flags = 0
      if block_index == 0 then
        flags = flags + CHUNK_START
      end
      local is_last_block = (block_index == num_blocks - 1)
      local is_last_chunk = (chunk_index == num_chunks - 1)
      if is_last_block then
        flags = flags + CHUNK_END
      end
      -- Root flag is set only on the final compression of a single-chunk
      -- input; multi-chunk inputs set ROOT during the parent-merge step.
      if is_last_block and is_last_chunk and num_chunks == 1 then
        flags = flags + ROOT
      end

      local words = words_from_block(block)
      local out = compress(cv, words, chunk_index, block_len, flags)
      cv = { out[1], out[2], out[3], out[4], out[5], out[6], out[7], out[8] }
    end
    chunk_cvs[#chunk_cvs + 1] = cv
  end

  -- Merge chunk chaining values up the tree. For the single-chunk case the
  -- ROOT flag was already applied above and the loop body is skipped.
  local function parent_cv(left, right, is_root)
    local block_words = {
      left[1], left[2], left[3], left[4], left[5], left[6], left[7], left[8],
      right[1], right[2], right[3], right[4], right[5], right[6], right[7], right[8],
    }
    local flags = 4 -- PARENT
    if is_root then
      flags = flags + ROOT
    end
    local out = compress(
      { IV[1], IV[2], IV[3], IV[4], IV[5], IV[6], IV[7], IV[8] },
      block_words, 0, BLOCK_LEN, flags
    )
    return { out[1], out[2], out[3], out[4], out[5], out[6], out[7], out[8] }
  end

  local root_words
  if num_chunks == 1 then
    -- The single chunk's final compression already had ROOT set, so the
    -- chaining value words are the root output words.
    root_words = chunk_cvs[1]
  else
    -- Reduce left-to-right. BLAKE3's tree is left-balanced: combine the
    -- running left subtree with each subsequent chunk CV. The final merge
    -- carries ROOT.
    local node = chunk_cvs[1]
    for i = 2, #chunk_cvs do
      node = parent_cv(node, chunk_cvs[i], i == #chunk_cvs)
    end
    root_words = node
  end

  -- Serialize the 8 root words little-endian into 32 bytes.
  local bytes = {}
  for i = 1, 8 do
    local w = root_words[i]
    bytes[#bytes + 1] = string.char(band(w, 0xFF))
    bytes[#bytes + 1] = string.char(band(rshift(w, 8), 0xFF))
    bytes[#bytes + 1] = string.char(band(rshift(w, 16), 0xFF))
    bytes[#bytes + 1] = string.char(band(rshift(w, 24), 0xFF))
  end
  return table.concat(bytes)
end

M.OUT_LEN = OUT_LEN
M.BLOCK_LEN = BLOCK_LEN
M.CHUNK_LEN = CHUNK_LEN
-- Silence unused-local lints for symbols kept for spec readability.
M._unused = { bnot, tobit }

return M
